raft.go

```diff
@@ -1479,11 +1488,32 @@ func (r *Raft) requestVote(rpc RPC, req *RequestVoteRequest) {
 	// check the LeadershipTransfer flag is set. Usually votes are rejected if
 	// there is a known leader. But if the leader initiated a leadership transfer,
 	// vote!
-	candidate := r.trans.DecodePeer(req.Candidate)
-	if leader := r.Leader(); leader != "" && leader != candidate && !req.LeadershipTransfer {
+	var candidate ServerAddress
+	var candidateBytes []byte
+	if req.ProtocolVersion > 3 {
+		candidate = r.trans.DecodePeer(req.Addr)
+		candidateBytes = req.Addr
+	} else {
+		candidate = r.trans.DecodePeer(req.Candidate)
+		candidateBytes = req.Candidate
```

Comment:
Similar comment here about if we need to do this. If we are still going to populate `req.Candidate` then maybe we keep using that field here?

Also similar comment about godoc. We should add a godoc to the `Candidate` field explaining whatever we decide (either that it's ignored in V4+, or that we continue to use it instead of the `RPCHeader.Addr`)

Reply:
See my comment about removing this at some point. The reason we are keeping it is for backward compatibility but we would like to remove this at some point.